text/0000-debugger-visualizer.md

```diff
@@ -0,0 +1,643 @@
+        </Synthetic>
+        <Synthetic Name="UpperRight">
+          <DisplayString>({x + dx}, {y + dy})</DisplayString>
+        </Synthetic>
+        <Synthetic Name="LowerRight">
+          <DisplayString>({x + dx}, {y})</DisplayString>
+        </Synthetic>
+      </Expand>
+    </Type>
+</AutoVisualizer>
+```
+
+When viewed under WinDbg, the `fancy_rect` variable would be shown as follows:
+
+```text
+> Variables:
+  > fancy_rect: (10, 10) + (5, 5)
+    > LowerLeft: (10, 10)
+    > UpperLeft: (10, 15)
+    > UpperRight: (15, 15)
+    > LowerRight: (15, 10)
+```
+
+## Supporting Pretty Printers
+
+This section describes how GDB's pretty printers are supported in Rust.
+
+To use a pretty printer, developers write python scripts that describe how a type
+should be displayed when loaded up in GDB/LLDB. (See: https://sourceware.org/gdb/onlinedocs/gdb/Pretty-Printing.html#Pretty-Printing)
+The pretty printers provide patterns, which match type names, and for matching
+types, descibe how to display those types. (For writing a pretty printer, see: https://sourceware.org/gdb/onlinedocs/gdb/Writing-a-Pretty_002dPrinter.html#Writing-a-Pretty_002dPrinter).
+
+Rust developers can add one or more pretty printers to their crate. This is done
+in the Rust compiler via `.py` python scripts. Through the use of a new Rust attribute,
+`#[debugger_visualizer]`, the compiler will encode the contents of the `.py` file in
+the crate metadata if the target is an `rlib`. If the target is an executable, the
+`.debug_gdb_scripts` section will include a reference to the pretty printer specified.
+
+To provide pretty printers, developers create a file with the `.py` file
+extension and reference it via the `#[debugger_visualizer]` attribute as follows:
+
+```rust
+#[debugger_visualizer(gdb_script_file = "../foo.py")]
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+In rustc, a new built-in attribute `#[debugger_visualizer]` will be added which
+instructs the compiler to take the specified file path for a debugger visualizer
+and add it to the current binary being built. The file path specified must be
+relative to the location of the attribute.
+
+The `#[debugger_visualizer]` attribute will reserve multiple keys to be able to
+specify which type of visualizer is being applied. The following keys will be
+reserved as part of this RFC:
+
+* `natvis_file`
+* `gdb_script_file`
+
+As more visualizer schemes arise, more keys may be added in the future to ensure
+a great debugging experience for any debugger that the Rust community sees fit.
+
+For example, to specify that a `.natvis` file should be included in the binary
+being built, the following attribute should be added to the Rust source:
+
+```rust
+#[debugger_visualizer(natvis_file = "../foo.natvis")]
+```
+
+The same can be done to specify a GDB python debugger script:
+
+```rust
+#[debugger_visualizer(gdb_script_file = "../foo.py")]
```

Comment:
Total nit but:

```suggestion
#[debugger_visualization(gdb_script_file = "../foo.py")]
```

maybe? (Unlike @joshtriplett, I would prefer full words, and not "debug_vis".)

I don't care to block on this detail either way. =)

Reply:
I agree, my preference would be to have the attribute contain full words. I think I'll leave the name as is for now since it sounds pretty clear. If there is a reason to change it to `#[debugger_visualization]` instead I would have no problem making that update.